File: src/main/java/com/forgeryclient/utilities/utils/FileUtils.java
package com.forgeryclient.utilities.utils;

import java.io.File;
import java.io.FilenameFilter;
import java.util.Objects;

public class FileUtils {

    public static void foreachDeep(File directory, FilenameFilter filter, FileAction action) {
        for (File f : Objects.requireNonNull(directory.listFiles(filter))) {
            if (f.isDirectory()) {
                foreachDeep(f, filter, action);
            } else {
                action.run(f);
            }
        }
    }

    public interface FileAction {
        void run(File f);
    }

}

File: src/main/kotlin/org/bundleproject/bundle/utils/FileUtils.kt
package org.bundleproject.bundle.utils

import java.io.File
import java.io.FileNotFoundException

object FileUtils {

    fun foreachFileDeep(rootDir: File, current: MutableList<File> = mutableListOf()): MutableList<File> {
        if (!rootDir.isDirectory) throw FileNotFoundException("File is not a directory!")

        for (file in rootDir.listFiles()!!) {
            if (file.isDirectory) foreachFileDeep(file, current)
            else current.add(file)
        }

        return current
    }

}